### PR TITLE
Implement servlet to create or update user lists

### PR DESCRIPTION
**Pull request by AnudeepYakkala**
Added servlet that handles a POST request to "/api/v1/create-or-update-user-list-servlet". This body of the POST request is parsed using the gson library. The writeUserlists() function is called from the SpannerUtilFunctions wrapper class to create or update the specified user list in the database. An empty response is returned with the appropriate status code. 

Added two tests using the Mockito library. One test checks to see if a proper request returns the right response and another tests checks to see if a bad request returns the right response. 

Also added a test axios POST request on the React side. 

Sorry that there are a lot of commits. I originally had the database specific functions but they were moved to SpannerUtilFunctions.

### CODE REVIEW

File: capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  private class UserList {
+    private int listId;
+    private String displayName;
+    private List<String> itemTypes;
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Gson g = new Gson();
+    SpannerUtilFunctions suf = new SpannerUtilFunctions();
```
Comment (chen-dawn) [inline capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java:36]: This is not implemented yet right?

File: capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
+package com.google.sps.servlets;
```
Comment (chen-dawn) [inline capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java:1]: do you mind moving this servlet to `java/com/google/servlets` directory? So the package is `com.google.servlets`. (delete the sps directory, since we don't need it).

File: capstone/client/src/components/ListPage.js [hunk 3]
```diff
@@ -74,6 +76,17 @@ class ListPage extends Component {
         alert: (<Alert severity="error">Please select at least one item!</Alert>),
       })
     } else {
+      const response = axios.post(
+        '/api/v1/create-or-update-user-list-servlet',
+        { 
+          userId: 1,
+          userList: {
+            listId: 1,
```
Comment (chen-dawn) [inline capstone/client/src/components/ListPage.js:84]: When the user first creates a list, how is the UI going to know what's the listId? Or how can it create a listId so it's unique? it's often better to assign the listId generation to the backend, so in your servlet (or database utils) you can have some helper function that generates a unique ID, and then maybe return the ID to the UI for subsequent operations. One possible unique ID is the userId + timestamp, for example. Or you can ask the DB to generate a unique key for you.

Comment (chen-dawn) [reply capstone/client/src/components/ListPage.js:84]: (maybe timestamp is not great because it's always increasing and will generate hotspots)
https://cloud.google.com/spanner/docs/schema-and-data-model#choosing_a_primary_key

File: capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java (hunk 1 above)
Comment (AnudeepYakkala) [reply capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java:36]: It is sort of implemented. I pushed the file to the DatabaseWrapperFunctions branch when I moved the database specific functions out of the servlets into a separate class since I already wrote the functions. I wasn't sure if Brett wanted to use what I had or change it so I didn't create a pull request.

File: capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java [hunk 4]
```diff
@@ -37,13 +33,23 @@ public ResponseBody(ResponseStatus responseStatus) {
...
+   } catch (IOException) {
+      response.setStatus(BAD_REQUEST)
+      response.getWriter().println("");
+      return;
+   }
+   try {
+     suf.writeUserLists(requestBody.userId. requestBody.UserList.listId, requestBody.UserList.itemTypes);
+   } catch (Exception e) {
+     response.setStatus(INTERNAL_SERVER_ERROR);
+     response.getWriter().println("");
```
Comment (chen-dawn) [inline capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java]: The error message should be descriptive

File: capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java (hunk 2 above)
Comment (AnudeepYakkala) [reply capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java:1]: Yeah, I will move them.

File: capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java [hunk 5]
```diff
@@ -37,13 +33,23 @@ public ResponseBody(ResponseStatus responseStatus) {
...
+   try {
+     RequestBody requestBody = getRequestBody(request);
+   } catch (IOException) {
+      response.setStatus(BAD_REQUEST)
+      response.getWriter().println("");
+      return;
+   }
+   try {
+     suf.writeUserLists(requestBody.userId. requestBody.UserList.listId, requestBody.UserList.itemTypes);
+   } catch (Exception e) {
```
Comment (chen-dawn) [inline capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java]: Try to figure out the more specific exception? also in your function declaration on line 34 you're only throwing IOException.

File: capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java (hunk 1 above)
Comment (chen-dawn) [reply capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java:36]: Ok, probably good to sync up with Brett on this.

Comment (AnudeepYakkala) [reply capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java:36]: Yeah, I messaged him about it.

File: capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java (hunk 5 above)
Comment (AnudeepYakkala) [reply capstone/backend/src/main/java/com/google/sps/servlets/CreateOrUpdateUserlistServlet.java]: I left that there because I didn't know what exception the database specific functions were going to throw. I changed it to check for the return value instead. This way the database wrapper functions can return false when there is an error.

File: capstone/backend/src/main/java/com/google/servlets/CreateOrUpdateUserlistServlet.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  private RequestBody getRequestBody(HttpServletRequest request) {
+    Gson g = new Gson();
+    String reqString = "";
+    String line;
+    try {
+      BufferedReader br = request.getReader();
+      while ((line = br.readLine()) != null) {
+        reqString += line;
```
Comment (luli0521) [inline capstone/backend/src/main/java/com/google/servlets/CreateOrUpdateUserlistServlet.java:56]: Use StringBuilder when concatenating strings in a loop for better performance.

File: capstone/backend/src/main/java/com/google/servlets/CreateOrUpdateUserlistServlet.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      return;
+    }
+    if (!suf.writeUserLists(
+            requestBody.userId, requestBody.userList.listId, requestBody.userList.itemTypes)) {
+      response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+      response.getWriter().println("An error occured while writing to the database.");
+      return;
+    }
+    response.setStatus(HttpServletResponse.SC_OK);
+    response.getWriter().println("");
```
Comment (luli0521) [inline capstone/backend/src/main/java/com/google/servlets/CreateOrUpdateUserlistServlet.java:46]: Shall we return the created/updated UserLists back in the response?

Comment (AnudeepYakkala) [reply capstone/backend/src/main/java/com/google/servlets/CreateOrUpdateUserlistServlet.java:46]: Yeah, I will change it so it returns the list.

File: capstone/backend/src/main/java/com/google/servlets/CreateOrUpdateUserlistServlet.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+      return;
+    }
+    ResponseBody responseBody = new ResponseBody(requestBody.userList);
+    response.setStatus(HttpServletResponse.SC_OK);
+    response.setContentType("application/json;");
+    response.getWriter().println(g.toJson(responseBody));
+  }
+
+  private long generateListId(int userId) {
+    long currentTime = System.currentTimeMillis();
```
Comment (chen-dawn) [inline capstone/backend/src/main/java/com/google/servlets/CreateOrUpdateUserlistServlet.java:64]: what do you plan to use for your userId hash function? if it's also timestamp based you'll have trouble here (timestamp is 10 digits or 13 digits, int64 only supports up to 19 digits)

File: capstone/backend/src/test/java/com/google/CreateOrUpdateUserlistServletTest.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+import org.mockito.Mockito;
+
+public class CreateOrUpdateUserlistServletTest extends TestCase {
+  public void testDoPostSucceed() throws ServletException, IOException {
+    HttpServletRequest req = Mockito.mock(HttpServletRequest.class);
+    HttpServletResponse res = Mockito.mock(HttpServletResponse.class);
+    StringWriter writer = new StringWriter();
+    StringReader reader = new StringReader(
+        "{\"userId\": 1,\"userList\":{\"listId\":2,\"displayName\":\"List Name\",\"itemTypes\":[\"Butter\",\"Orange Juice\", \"Peanuts\"]}");
+    CreateOrUpdateUserlistServlet servlet = new CreateOrUpdateUserlistServlet();
```
Comment (luli0521) [inline capstone/backend/src/test/java/com/google/CreateOrUpdateUserlistServletTest.java:20]: Move this line to just before 'servlet.doPost'.

Also usually we divide a test function into three parts, separated by an empty line:
1. The mock data setup.
2. The actual call.
3. The verification steps.

If you follow this pattern, you will see that in part 1, the only difference between this and the test below is the StringReader parameter, so you can extract a common function 'setupMockData' and pass the string as a parameter.

Similarly part 2 can also be extracted into a common function.

File: capstone/backend/src/main/java/com/google/servlets/CreateOrUpdateUserlistServlet.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    String id = String.valueOf(userId) + currentTime;
+    return Long.parseLong(id);
+  }
+
+  private RequestBody getRequestBody(HttpServletRequest request) {
```
Comment (chen-dawn) [inline capstone/backend/src/main/java/com/google/servlets/CreateOrUpdateUserlistServlet.java:69]: Personal preference: The top part is actually `getRequestBody`, the second part is more like a `requestValidator`. I would make a `String getRequestBody(HttpServletRequest request)` function that really can be shared across servlets for parsing the body, then have a seperate validator for the specific request.
